cpp/include/raft/matrix/detail/select_radix_updated.cuh

```diff
@@ -0,0 +1,1183 @@
+ * @tparam BitsPerPass
+ *   The size of the radix;
+ *   it affects the number of passes and number of buckets.
+ * @tparam BlockSize
+ *   Number of threads in a kernel thread block.
+ *
+ * @param[in] in
+ *   contiguous device array of inputs of size (len * batch_size);
+ *   these are compared and selected.
+ * @param[in] in_idx
+ *   contiguous device array of inputs of size (len * batch_size);
+ *   typically, these are indices of the corresponding in_keys.
+ * @param batch_size
+ *   number of input rows, i.e. the batch size.
+ * @param len
+ *   length of a single input array (row); also sometimes referred as n_cols.
+ *   Invariant: len >= k.
+ * @param k
+ *   the number of outputs to select in each input row.
+ * @param[out] out
+ *   contiguous device array of outputs of size (k * batch_size);
+ *   the k smallest/largest values from each row of the `in_keys`.
+ * @param[out] out_idx
+ *   contiguous device array of outputs of size (k * batch_size);
+ *   the payload selected together with `out`.
+ * @param select_min
+ *   whether to select k smallest (true) or largest (false) keys.
+ * @param fused_last_filter
+ *   when it's true, the last filter is fused into the kernel in the last pass and only one thread
+ *   block will do the filtering; when false, a standalone filter kernel with multiple thread
+ *   blocks is called. The later case is preferable when the most significant bits of input data are
+ *   almost the same. That is, when the value range of input data is narrow. In such case, there
+ *   could be a large number of inputs for the last filter, hence using multiple thread blocks is
+ *   beneficial.
+ * @param stream
+ * @param mr an optional memory resource to use across the calls (you can provide a large enough
+ *           memory pool here to avoid memory allocations within the call).
+ */
+template <typename T, typename IdxT, int BitsPerPass, int BlockSize>
+void select_k_updated(const T* in,
+                      const IdxT* in_idx,
+                      int batch_size,
+                      IdxT len,
+                      IdxT k,
+                      T* out,
+                      IdxT* out_idx,
+                      bool select_min,
+                      bool fused_last_filter,
+                      rmm::cuda_stream_view stream,
+                      rmm::mr::device_memory_resource* mr = nullptr)
+{
+  if (k == len) {
+    RAFT_CUDA_TRY(
+      cudaMemcpyAsync(out, in, sizeof(T) * batch_size * len, cudaMemcpyDeviceToDevice, stream));
+    if (in_idx) {
+      RAFT_CUDA_TRY(cudaMemcpyAsync(
+        out_idx, in_idx, sizeof(IdxT) * batch_size * len, cudaMemcpyDeviceToDevice, stream));
+    } else {
+      auto out_idx_view =
+        raft::make_device_vector_view(out_idx, static_cast<size_t>(len) * batch_size);
+      raft::device_resources handle(stream);
+      raft::linalg::map_offset(handle, out_idx_view, raft::mod_const_op<IdxT>(len));
+    }
+    return;
+  }
+
+  int sm_cnt;
+  {
+    int dev;
```

Comment:
Thanks for the suggestion.
I have a question about how to use it. The API seems to be [`device_resources::get_device_properties()`](https://github.com/rapidsai/raft/blob/d45eb1ec117bfa04e0af3b576fc2baada39462bb/cpp/include/raft/core/device_resources.hpp#L229). How should I get a `device_resources` object?
If it's defined locally inside `select_k_updated()`, the underline function `cudaGetDeviceProperties()` will be called every time when `select_k_updated()` is called. The overhead is too large: `cudaGetDeviceProperties()` takes more than 900us, while the whole top-k computation takes ~40us when batch size = 1 and len = 1e6.
It may be better to change the [`raft::matrix::detail::select_k`](https://github.com/rapidsai/raft/blob/d45eb1ec117bfa04e0af3b576fc2baada39462bb/cpp/include/raft/matrix/detail/select_k.cuh#L67) interface to add a `device_resources&` parameter? But it will require changing the declarations of all select-k implementations.

Reply:
Ah, you're right, we don't pass the `device_resources` in this api yet. Normally, we pass it from outside, so that the underline `cudaGetDeviceProperties` would be called only once. Maybe let's just add a TODO comment here until we do the refactoring of the whole `select_k`?  (otherwise we'd have to change the calling code in ivf_flat and ivf_pq, which is surely out of scope of this pr)

Reply:
Got it; we can refactor it in future. TODO comment is added.

Reply:
@yong-wang would you mind also creating a Github issue for the change just so it stays on our radar?

Reply:
Good idea. Issue #1294 is created.